src/operations/blas3/gemm_no_local.hpp

```diff
@@ -394,76 +415,453 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
+   * @tparam BoundaryCheckM : Type of function for checking boundary in M
+   * dimension.
+   * @tparam BoundaryCheckN : Type of function for checking boundary in N
+   * dimension.
+   * @tparam PointerType : Type of the input pointers for A and B matrices.
+   * @param A : Input matrix A
+   * @param B : Input matrix B
+   * @param boundary_check_m : Function which checks boundary in M dimension.
+   * @param boundary_check_n : Function which checks boundary in N dimension.
+   * @param A_ptr_index : index of next element for A
+   * @param B_ptr_index : index of next element for B
+   * @param lda : leading dimension of A
+   * @param ldb : leading dimension of B
+   * @param k : the current value of K from the main loop which calls this
+   * method.
+   * @param reg_a : Pointer to private register for A.
+   * @param reg_b : Pointer to private register for B.
+   * @param reg_res : Pointer to private register for partial result.
+   * @param out_of_range : Controls whether to exit some functions early if
+   * block is out of range of A or B.*/
+  template <index_t packet_size, bool check_boundary, bool check_k,
+            typename BoundaryCheckM, typename BoundaryCheckN,
+            typename PointerType>
+  SYCL_BLAS_INLINE void load_and_compute_block(
+      PointerType A, PointerType B, BoundaryCheckM boundary_check_m,
+      BoundaryCheckN boundary_check_n, const index_t &A_ptr_index,
+      const index_t &B_ptr_index, const index_t &lda, const index_t &ldb,
+      const index_t &k, element_t *reg_a, element_t *reg_b, element_t *reg_res,
+      bool out_of_range
+#ifdef ARM_GPU
+      ,
+      const cl::sycl::nd_item<1> &id
+#endif
+  ) {
+    /*
+     * Loading a corresponding block of matrix A into reg_a
+     */
+    load_block_a<item_rows, packet_size, wg_rows * packet_size, check_boundary,
+                 check_k, packet_size, trans_a>(
+        A, reg_a, A_ptr_index, lda, boundary_check_m,
+        [=](const index_t &idx) SYCL_BLAS_ALWAYS_INLINE { return idx < k; },
+        out_of_range);
+#ifdef ARM_GPU
+    id.barrier(cl::sycl::access::fence_space::local_space);
+#endif
+
+#pragma unroll
+    for (int j = 0; j < packet_size; j++) {
+#pragma unroll
+      for (int i = 0; i < item_cols / packet_size; i++) {
+        const index_t ofs = i * B_ptr_index;
+        const index_t col_ofs = i * (trans_b ? packet_size : 1) * wg_cols;
+
+        /*
+         * Loading a corresponding partial block of matrix B into reg_b
+         */
+        load_single_b<check_k, check_boundary, packet_size, trans_b>(
+            B + ofs, reg_b, j, col_ofs,
+            [=](const index_t &idx) SYCL_BLAS_ALWAYS_INLINE { return idx < k; },
+            boundary_check_n, out_of_range);
+
+        /*
+         * Computing a partial GEMM for the loaded block of reg_a and partial
+         * reg_b and adding the result into reg_res
+         */
+
+        compute_block_gemm_no_shared<packet_size>(i + j, reg_a, reg_b, reg_res);
```

Comment:
the iteration type is const Index& so (i + j) force compiler to allocate memory and pass the address of that memory to the function, There is no need to make an Index to be a reference

Reply:
Fixed!